hilti/toolchain/src/ast/node.cc

```diff
@@ -23,7 +23,10 @@ std::string node::to_string(const Tags& ti) {
     return util::join(util::transform(ti, [](auto i) { return std::to_string(i); }), ",");
 }
 
-Node::~Node() {}
+Node::~Node() {
+    clearChildren();
+    _ref_count = -1; // for debugging, mark as destroyed
```

Comment:
Suggest to drop this. The object with the updated `_ref_count` is going away so this update is shouldn't be observable outside of UB. We should also drop the checks for `_ref_count != -1` and could then make `_ref_count` unsigned.

```suggestion
```

Reply:
I'd like to keep this, it's helpful for debugging. Because we're mostly passing around raw pointers, it's hard to catch dangling pointers in the new scheme. When looking at a `Node*` in a debugger, there's no way to tell if that instance actually still exists (ref_cnt==zero is legitimate). Setting the count to minus one says it's been destroyed (of course it's not guaranteed that the memory remains set like that, but it's quite likely). Similar for the asserts: it has a good chance of catching memory mgmt errors.